WIP: Allow generating iconset for HASS icons

balloob opened the pull request:
# WIP until we migrate to Polymer 3

Included icons will be limited to our official icons (the ones we refer in the frontend code). Size of included icons will be reduced from 747K to 38K.

This new gulp script will:
 - Search through our code and find all icons that we use in our official code.
 - Create a new icon set called `hass` based on icon set `mdi` with only the icons that we use

To do:
 - [x] We should refer to icons in our code as `hass:` and update our search regex
 - [ ] ~~We should add a config option to the frontend component to allow loading the MDI icons~~
 - [ ] ~~We should also search the python code for used icons??~~
 - [x] Fork `iron-icon` to see if missing iconset is `mdi` and load `mdi` when missing
 - [x] Update gen-index-html script
 - [x] Convert download mdi script to a pure JS solution instead of a Python/JS hybrid
 - [x] Get it to work. Somehow loading mdi on demand does not seem to work

Not part of this PR but would help loading time:
 - [ ] Migrate as much Python code in the backend to use device_class instead of icon.

balloob committed:
Allow generating iconset for HASS icons

andrey-git commented:
Yes, we should also search for python code.

However this makes different namespace difficult

NovapaX commented:
We could fork `iron-icon` or `iron-iconset` and let us provide two sources for one iconset. If the primary source (i.e. our packed `mdi:` subset) fails, load the fallback and try there....
Although a typo in an icon-name would immediately load the fallback iconset.

balloob commented:
What if we get rid of all `icon` definitions on the Python side as part of the core. Instead, icons should be based on `device_class` or be derived from domain. That way we don't have to search the Python code.

balloob commented:
Well, removing icons from the backend might be quite the task…

![image](https://user-images.githubusercontent.com/1444314/39026501-7f10c0fc-441b-11e8-85c0-7bf297862c0d.png)

balloob commented:
Guess defining an icon in the backed can sometimes make sense, like with Alpha Advantage:

![image](https://user-images.githubusercontent.com/1444314/39026530-a0462e2e-441b-11e8-84a6-a20f58467897.png)

balloob commented:
Probably best approach would be to patch `iron-icon` to trigger it to load `mdi` icon set.

We would patch around here:

https://github.com/PolymerElements/iron-icon/blob/7f735696e8d8f6edddb16d78da7da94a3638fa08/iron-icon.html#L182-L184

balloob committed:
Lint

balloob committed:
Load mdi on demand

balloob committed:
mdi -> hass

balloob committed:
Lint

balloob commented:
So I played a bit more with this. Added a new `ha-icon` which is a fork of `iron-icon` but with the code attached to load `mdi.html` on demand. However, this doesn't seem to work 🤔 

![image](https://user-images.githubusercontent.com/1444314/39029435-ccfa4424-4429-11e8-9fde-418557130f95.png)

The dynamic panels still use mdi icons. As you can see, icon for Configuration is missing. I have confirmed in the network panel that `mdi.html` has loaded, yet the iconset doesn't seem to get registered.

balloob commented:
That's totally it. When I patch `iron-iconset-svg` to use the PR by @andrey-git it works.

balloob committed:
Hack reschedule

balloob commented:
Since I'm already updating the `updateIcon` method of `ha-icon` to add `mdi` on-demand loading, I also introduced a little hack to detect if an iconset might not be loading yet:

b271518df45fc334e3c82a33850d1e54ca7dabae

It's not pretty.

balloob committed:
Revert mdi -> hass change for customize

balloob committed:
Clarify on top of ha-icon what has changed

balloob committed:
Convert latest mdi icon references

balloob committed:
Convert python script to JS

balloob committed:
BUILD_DEV=0 all the things

balloob committed:
Embed built-in panel icons

balloob committed:
service worker hass icons fix

balloob committed:
Lint

balloob commented:
I've converted the MDI download script to JavaScript, dropping the last Python dependency 🎉 

PR on backend to convert built-in panels to use HASS iconset: https://github.com/home-assistant/home-assistant/pull/14185

built-in panel icons are hardcoded in the gulp task.

Updated service worker to handle fingerprinting of `hass_icons.html` correctly and handle the lack of fingerprinting on `mdi.html` correctly.

andrey-git commented:
(didn't look at the code yet) I think we should support icons set by platforms (in Python)

balloob commented:
Problem with icon set per platform is the duplication that might occur when we introduce even more namespaces. We currently have overlap between hass and mdi namespace, but we're assuming that the mdi namespace might not be necessary for basic users.

I prefer if we first see how far sensor device classes is going to get us before we invest time in making icon sets for Python code. In a perfect world, the Python code should not be aware of icons, although I do realize that it's better for us to break that rule here to make the simple things not overly complicated.

c727 commented on gulp/tasks/gen-service-worker.js:
```diff
@@ -23,7 +23,7 @@ const DEV = !!JSON.parse(process.env.BUILD_DEV || 'true');
 const dynamicUrlToDependencies = {};
 
 const staticFingerprinted = [
-  'mdi.html',
```
we still have to fingerprint mdi

balloob commented:
This has been fixed in #1214.
